### ISSUE TITLE
Statement doesn't timeout when the connection is established but is actually broken.

### ISSUE THREAD

**Issue by memosstilvi** [rails/rails#22557]
We currently use ActiveRecord to query an RDS instance from our servers that run on EC2. 

When the RDS instance reboots with failover (which is not often but might happen under heavy load) what AWS does is to switch the DNS to point to a back up instance while restarting the primary one.

This causes the connections from the servers to the database to remain `ESTABLISHED` (as shown by `netstats` output), without working however.

When doing normal Reboot (without failover) the connections change to `CLOSE_WAIT` which allows us to catch the exception in the application level and reconnect.

However in the `Reboot with Failover` case, the request to the database keeps hanging indefinitely, causing subsequent requests to pile up and eventually kill our servers.

I played with the timeout parameters in `database.yml` but nothing seems to prevent the requests from hanging. I also tried to wrap `Timeout::timeout` around our DB calls, but this didn't solve the issue either. (probably because the `Timeout::Error` exception is caught somewhere else)

I would appreciate any advice if anyone has experienced the same issue before or know how to solve it.

**Comment by jgross206** [comment]
+1 I am having this exact issue as well.  I've found that *eventually* (like 10-15 minutes later) the request will raise an exception.  Please let me know if anyone has a possible fix or workaround.

**Comment by jgross206** [comment]
@memosstilvi Seems like this is a TCP issue.  Since the server we were connected to is completely gone, there's no RST response and we just keep trying with exponential(ish) backoff which takes ~15 minutes before giving up (at least on my system).

If you reduce the kernel-level tcp_retries2 parameter it fails much more quickly:
    sudo sysctl -w net.ipv4.tcp_retries2=5

Not saying that this is a good, permanent, or robust solution but hopefully it will help you with your own troubleshooting.  Ideally there will be a way to tune these params only within ActiveRecord, but I haven't figured out how to do that yet.

**Comment by aquach** [comment]
I ran into this issue yesterday and seemed to get something working.

For context, I'm using the mysql2 gem, which has two relevant timeout parameters: connect_timeout, for initial connect, and read_timeout, for reading a response. I'm not sure which one you're using, but these were relevant to my findings.

With mysql2, I was able to failover with minimal latency by setting the read_timeout and connect_timeout appropriately low and enabling reconnect: true. With read_timeout, Mysql2 will abort and close the connection, which means that you don't have to wait for the TCP retransmission timeout. With reconnect: true, Mysql2 will attempt to reestablish closed connections, which is obviously necessary for failover to the new server. And with connect_timeout, you ensure that Ruby doesn't block for a long time waiting for a response to a SYN that will never arrive because the failover server hasn't gotten online yet.

Here's the behavior I observed with these settings:

```
read_timeout: 20
connect_timeout: 30
reconnect: true
```

I kept this script running through the failover operation to test queries:

```
loop do
  sleep 1
  begin
    ActiveRecord::Base.connection.execute 'show tables'
  rescue Exception => e
    p e
  end
end
```

With these results:

```
D, [2016-02-02T18:39:03.505834 #20388] DEBUG -- :    (1.3ms)  show tables
D, [2016-02-02T18:39:04.507482 #20388] DEBUG -- :    (1.3ms)  show tables
D, [2016-02-02T18:39:05.509194 #20388] DEBUG -- :    (1.4ms)  show tables
D, [2016-02-02T18:39:06.510846 #20388] DEBUG -- :    (1.3ms)  show tables
D, [2016-02-02T18:39:07.512450 #20388] DEBUG -- :    (1.3ms)  show tables
D, [2016-02-02T18:39:08.514074 #20388] DEBUG -- :    (1.3ms)  show tables
D, [2016-02-02T18:39:09.515776 #20388] DEBUG -- :    (1.3ms)  show tables
D, [2016-02-02T18:39:10.517407 #20388] DEBUG -- :    (1.3ms)  show tables
D, [2016-02-02T18:39:11.519070 #20388] DEBUG -- :    (1.3ms)  show tables
D, [2016-02-02T18:39:12.520785 #20388] DEBUG -- :    (1.3ms)  show tables
D, [2016-02-02T18:39:13.522382 #20388] DEBUG -- :    (1.2ms)  show tables
D, [2016-02-02T18:39:14.523909 #20388] DEBUG -- :    (1.2ms)  show tables
D, [2016-02-02T18:39:35.542818 #20388] DEBUG -- :    (20018.6ms)  show tables
#<ActiveRecord::StatementInvalid: Mysql2::Error: Timeout waiting for a query. (waited 20 seconds): show tables>
D, [2016-02-02T18:40:06.584210 #20388] DEBUG -- :    (30040.9ms)  show tables
#<ActiveRecord::StatementInvalid: Mysql2::Error: Can't connect to MySQL 2ujpxm.us-west-2.rds.amazonaws.com' (4): show tables>
D, [2016-02-02T18:40:07.589584 #20388] DEBUG -- :    (4.9ms)  show tables
D, [2016-02-02T18:40:08.590630 #20388] DEBUG -- :    (0.7ms)  show tables
D, [2016-02-02T18:40:09.591687 #20388] DEBUG -- :    (0.7ms)  show tables
```

You can see why those three parameters are necessary to effect a failover; otherwise the connection would just hang without moving forward from timeout -> close -> reconnect.

By the way, the DNS swap from old to new server occurred at 18:39:55, 20 seconds after the first query hung. I'm not sure if that 20 seconds is related to the 20 second read timeout though.

In practice, I think it's best to have 2-5 second read and connect timeouts, which will minimize the ephemeral hung time during the failover.

In summary, I was able to use mysql2 timeouts to ensure that the server was always progressing and attempting to reconnect to the new server.

**Comment by fedeest** [comment]
Hi, sory for reviving this after to long, but I've been having this same issue and doing what @jgross206 said:
`sudo sysctl -w net.ipv4.tcp_retries2=5`
Worked quite well. But, well, this is not something really clean and nice. Is there any way to set this in the ActiveRecord options or make it specific por the application and not system wide?
